packages/compass-user-data/src/user-data.ts

```diff
@@ -0,0 +1,185 @@
+  basePath?: string;
+  serialize?: SerializeContent<T>;
+  deserialize?: DeserializeContent<T>;
+};
+
+type ReadOptions = {
+  ignoreErrors: boolean;
+};
+
+export class UserData<T> {
+  private readonly subdir: string;
+  private readonly basePath?: string;
+  private readonly serialize: SerializeContent<T>;
+  private readonly deserialize: DeserializeContent<T>;
+
+  constructor({
+    subdir,
+    basePath,
+    serialize = (content: T) => JSON.stringify(content, null, 2),
+    deserialize = JSON.parse,
+  }: UserDataOptions<T>) {
+    this.subdir = subdir;
+    this.basePath = basePath;
+    this.deserialize = deserialize;
+    this.serialize = serialize;
+  }
+
+  private async getEnsuredBasePath(): Promise<string> {
+    const basepath = this.basePath ? this.basePath : getStoragePath();
+
+    const root = path.join(basepath, this.subdir);
+
+    await fs.mkdir(root, { recursive: true });
+
+    return root;
+  }
+
+  private async getFileAbsolutePath(filepath?: string): Promise<string> {
+    const root = await this.getEnsuredBasePath();
+    const pathRelativeToRoot = path.relative(
+      root,
+      path.join(root, filepath ?? '')
+    );
+
+    if (
+      pathRelativeToRoot.startsWith('..') ||
+      path.isAbsolute(pathRelativeToRoot)
+    ) {
+      throw new Error(
+        `Invalid file path: '${filepath}' is not a subpath of ${root}.`
+      );
+    }
+
+    return path.resolve(root, pathRelativeToRoot);
+  }
+
+  private async readAndParseFile(absolutePath: string, options: ReadOptions) {
+    let data: string;
+    try {
+      data = await fs.readFile(absolutePath, 'utf-8');
+    } catch (error) {
+      log.error(mongoLogId(1_001_000_234), 'Filesystem', 'Error reading file', {
+        path: absolutePath,
+        error: (error as Error).message,
+      });
+      if (options.ignoreErrors) {
+        return undefined;
+      }
+      throw error;
+    }
+
+    try {
+      return this.deserialize(data);
+    } catch (error) {
+      log.error(mongoLogId(1_001_000_235), 'Filesystem', 'Error parsing data', {
+        path: absolutePath,
+        error: (error as Error).message,
+      });
+      if (options.ignoreErrors) {
+        return undefined;
+      }
+      throw error;
+    }
+  }
+
+  async readAll(
+    options: ReadOptions = {
+      ignoreErrors: true,
+    }
+  ) {
+    const absolutePath = await this.getFileAbsolutePath();
+    const filePathList = await fs.readdir(absolutePath);
+    const data = await Promise.allSettled(
+      filePathList.map((x) => this.readOne(x, options))
+    );
+
+    const result = {
+      data: [],
+      errors: [],
+    } as {
+      data: T[];
+      errors: Error[];
+    };
+
+    for (const item of data) {
+      if (item.status === 'fulfilled' && item.value) {
+        result.data.push(item.value);
+      }
+      if (item.status === 'rejected') {
+        result.errors.push(item.reason);
+      }
+    }
+
+    return result;
+  }
+
+  async readOne(
+    filepath: string,
```

Comment:
I see that every place where we use this, a parent class that uses `UserData` has to implement some sort of `getFileName` method where `.json` extension is stuck at the end of some sort of id, would it make sense to change the interface here to encapsulate this logic inside `UserData`?